graphql_compiler/schema_transformation/split_query.py

```diff
@@ -559,3 +726,54 @@ def assign_and_return_out_name(self):
         self.intermediate_output_count += 1
         self.intermediate_output_names.add(out_name)
         return out_name
+
+
+class SchemaIdSetterVisitor(Visitor):
+    def __init__(self, type_info, query_node, type_name_to_schema_id):
+        """Create a visitor for setting the schema_id of the input query node.
+
+        Args:
+            type_info: TypeInfo, used to keep track of types of fields while traversing the AST
+            query_node: SubQueryNode, whose schema_id will be modified
+            type_name_to_schema_id: Dict[str, str], mapping the names of types to the id of the
+                                    schema that they came from
+        """
+        self.type_info = type_info
+        self.query_node = query_node
+        self.type_name_to_schema_id = type_name_to_schema_id
+
+    def enter_Field(self, *args):
+        """Check the schema of the type that the field leads to"""
+        child_type_name = strip_non_null_and_list_from_type(self.type_info.get_type()).name
+        self._check_or_set_schema_id(child_type_name)
+
+    def enter_InlineFragment(self, node, *args):
+        """Check the schema of the coerced type."""
+        self._check_or_set_schema_id(node.type_condition.name.value)
+
+    def _check_or_set_schema_id(self, type_name):
+        """Set the schema id of the root node if not yet set, otherwise check schema ids agree.
+
+        Args:
+            type_name: str, name of the type whose schema id we're comparing against the
+                       previously recorded schema id
+        """
+        if type_name in self.type_name_to_schema_id:  # It may be a scalar, and thus not found
+            current_type_schema_id = self.type_name_to_schema_id[type_name]
+            prior_type_schema_id = self.query_node.schema_id
+            if prior_type_schema_id is None:  # First time checking schema_id
+                self.query_node.schema_id = current_type_schema_id
+            elif current_type_schema_id != prior_type_schema_id:
+                # A single query piece has types from two schemas -- merged_schema_descriptor
+                # is invalid: an edge field without a @stitch directive crosses schemas,
+                # or type_name_to_schema_id is wrong
+                raise SchemaStructureError(
```

Comment:
Very nice! Would it be a lot of work to figure out one example of such a type that came in from a different schema without a stitch? If it's a lot of work then don't worry about it, but if it's relatively simple I think it would make a scary user-facing error seem quite a bit less scary.

Reply:
I think it's not hard to write a validator where you pass in a schema AST and a map of type name to schema id, and the validator spits out fields that cross schemas but don't have a `@stitch`, if any. I don't think it's easy to do right here though with the information we have.